homeassistant/components/voice_assistant/pipeline.py

```diff
@@ -169,47 +337,65 @@ async def execute(
             timeout=timeout,
         )
 
-    @abstractmethod
     async def _execute(self, run: PipelineRun):
-        """Run pipeline with request info and context."""
-
-
-@dataclass
-class TextPipelineRequest(PipelineRequest):
-    """Request to run the text portion only of a pipeline."""
-
-    intent_input: str
-    conversation_id: str | None = None
+        self._validate(run.start_stage)
 
-    async def _execute(
-        self,
-        run: PipelineRun,
-    ):
+        # stt -> intent -> tts
         run.start()
-        await run.recognize_intent(self.intent_input, self.conversation_id)
-        run.finish()
-
-
-@dataclass
-class AudioPipelineRequest(PipelineRequest):
-    """Request to full pipeline from audio input (stt) to audio output (tts)."""
-
-    intent_input: str  # this will be changed to stt audio
-    conversation_id: str | None = None
-
-    async def _execute(self, run: PipelineRun):
-        run.start()
-
-        # stt will go here
+        current_stage = run.start_stage
+
+        # Speech to text
+        intent_input = self.intent_input
+        if current_stage == PipelineStage.STT:
+            assert self.binary_handler_id is not None
+            assert self.stt_metadata is not None
+            assert self.stt_stream is not None
+            intent_input = await run.speech_to_text(
+                self.stt_metadata, self.stt_stream, self.binary_handler_id
+            )
+            current_stage = PipelineStage.INTENT
 
-        conversation_result = await run.recognize_intent(
-            self.intent_input, self.conversation_id
-        )
+        if run.end_stage != PipelineStage.STT:
+            tts_input = self.tts_input
 
-        tts_input = conversation_result.response.speech.get("plain", {}).get(
-            "speech", ""
-        )
+            if current_stage == PipelineStage.INTENT:
+                assert intent_input is not None
+                tts_input = await run.recognize_intent(
+                    intent_input, self.conversation_id
+                )
+                current_stage = PipelineStage.TTS
 
-        await run.text_to_speech(tts_input)
+            if run.end_stage != PipelineStage.INTENT:
+                if current_stage == PipelineStage.TTS:
+                    assert tts_input is not None
+                    await run.text_to_speech(tts_input)
 
         run.finish()
+
+    def _validate(self, stage: PipelineStage):
+        """Validate pipeline input against start stage."""
+        if stage == PipelineStage.STT:
+            if self.binary_handler_id is None:
+                raise PipelineRunValidationError(
```

Comment:
Should we raise or send an error event? We need to translate this to an event anyway for the websocket.

Reply:
This feels wrong in general, as details of the websocket API are bleeding into the pipeline. I think the handler id should be returned in a separate message before the pipeline starts.